README.markdown
# countable

A jQuery plugin that adds a character counter to inputs and textareas.

This is a slight modification of Brandon Aaron's work to support disabling form submit buttons when the maxLength is surpassed.

## Settings

The countable plugin has 15 settings:

* `threshold` - The percentage at which the counter begins to fade in. Default is **0.5**.
* `appendMethod` - One of **insertAfter**, **insertBefore**, **prependTo**, or **appendTo** to insert the counter after or before the input/textarea or prepend or append to the **target** element. Default is **'insertAfter'**.
* `target` - The element used relatively for the `appendMethod` option. Default is the actual input or textarea.
* `startOpacity` - The percentage of opacity it should start out with once it reaches the threshold. Default is **0.25**.
* `maxLength` - The maximum number of characters. Default uses the **maxlength** attribute of the input/textarea.
* `maxClassName` - The class name to add once the user has gone over the max number of characters. Default is **'maxed'**.
* `className` - The class name for the counter. Default is **'counter'**.
* `tagName` - The type of tag to use for the counter. Default is `span`.
* `submitButton` - The input element(s) that should be disabled when the maxLength is surpassed. Default is all submit buttons in the current form. Pass **''** to disable this functionality.
* `submitDisabledClassName` - The class to apply to the submitButton when it is disabled. Default is **'disabled'**
* `interval` - The interval at which it checks the input/textarea. Default is **750**.
* `positiveCopy` - The copy to use when the character count is below the max. Use `{n}` to denote where the number should go. Default is **'You have {n} characters left.'**.
* `negativeCopy` - The copy to use when the character count is over the max. Use `{n}` to denote where the number should go. Default is **'You are {n} characters over.'**.
* `fadeDuration` - The duration of the fade animations. Default is **'normal'**.
* `defaultText` - Text to ignore in the character count. This would be used if you have default text that disappears when focus is brought to the text box. Default is **''**.

## Examples

Add counter to all textarea objects on the page using defaults:

    $("textarea").countable();

Add counters to a specific set of text fields, before each element:

    $("#myTextBox, textarea.info").countable({ appendMethod: "insertBefore" });

Add counter to a specific textarea, but display in a given container element (with other advanced options):

    $("#comments").each(function() {
      $(this).countable({
        threshold: .75,
        appendMethod: "appendTo",
        target: $(this).parent().find("label small"),
        startOpacity: 1,
        maxLength: $(this).metadata().maxLength,
        submitButton: $(this).closest("form").find("input[type=submit]");
        positiveCopy: " ({n} characters left)",
        negativeCopy: " ({n} characters over the limit)",
        fadeDuration: 0,
        defaultText: "Enter message"
      })
    });

_Note: In the last example, the code is wrapped in an `each()` function call to get `this` to point to the text element. Also, `maxLength` is being pulled from contextual information using the metadata jQuery plugin (<http://plugins.jquery.com/project/metadata>)._

## License

The countable plugin is licensed under the MIT License (LICENSE.txt).

Copyright (c) 2010 [Brandon Aaron](http://brandonaaron.net)

## Contributors

Neil Monroe (neil.monroe[at]gmail.com)
Mina Mikhail (mina[at]fightingtheboss.com)

CHANGELOG.markdown
# countable Changelog

## 2010-06-01
Forked the original project from Brandon Aaron; Added parameters for disabling a form submit button when the maxCount is reached.

## 2010-02-11
Releasing as 1.0 :)

## 2009-08-28
Added 'defaultText' option to specify text to ignore if using label text inside the field that is not considered user-entered content. *By Neil Monroe (neil.monroe[at]gmail.com)*

## 2009-03-20
Enhanced Brandon's version by adding the ability to apply the count to a target container. *By Neil Monroe (neil.monroe[at]gmail.com)*


tests/test.js

test("input", 8, function() {
    var $el = $('#test1').countable(),
        $counter = $el.next();
    
    ok($counter.is('span'), 'counter is a span tag');
    ok($counter.is('.counter'), 'counter has correct class name');
    ok($counter.is(':hidden'), 'counter should be hidden initially');
    
    var str = '';
    for (var i=0; i<13; i++) str+='1';
    $el.val(str);
    $el.trigger('keyup');
    
    ok($counter.is(':visible'), 'counter should be visible since threshold has been met');
    ok($counter.find('.num').text() == "12", 'number of characters left should be 12');
    
    var str = '';
    for (var i=0; i<24; i++) str+='1';
    $el.val(str);
    $el.trigger('keyup');
    
    ok($counter.is(':visible'), 'counter should be visible');
    ok($counter.find('.num').text() == "1", 'number of characters left should be 1');
    
    var str = '';
    for (var i=0; i<5; i++) str+='1';
    $el.val(str);
    $el.trigger('keyup');
    
    ok($counter.is(':hidden'), 'counter should be hidden since threshold hasn\'t been met');
});

test("textarea", 8, function() {
    var $el = $('#test2').countable(),
        $counter = $el.next();
    
    ok($counter.is('span'), 'counter is a span tag');
    ok($counter.is('.counter'), 'counter has correct class name');
    ok($counter.is(':hidden'), 'counter should be hidden initially');
    
    var str = '';
    for (var i=0; i<13; i++) str+='1';
    $el.val(str);
    $el.trigger('keyup');
    
    ok($counter.is(':visible'), 'counter should be visible since threshold has been met');
    ok($counter.find('.num').text() == "12", 'number of characters left should be 12');
    
    var str = '';
    for (var i=0; i<24; i++) str+='1';
    $el.val(str);
    $el.trigger('keyup');
    
    ok($counter.is(':visible'), 'counter should be visible');
    ok($counter.find('.num').text() == "1", 'number of characters left should be 1');
    
    var str = '';
    for (var i=0; i<5; i++) str+='1';
    $el.val(str);
    $el.trigger('keyup');
    
    ok($counter.is(':hidden'), 'counter should be hidden since threshold hasn\'t been met');
});

jquery.countable.js
/*! Copyright (c) 2010 Brandon Aaron (http://brandonaaron.net)
 * Licensed under the MIT License (LICENSE.txt).
 *
 * Version 1.0
 *
 * Contributions by:
 *   - Neil Monroe (neil.monroe[at]gmail.com)
 *
 * Modified by:
*    - Mina Mikhail (mina.mikhail[at]gmail.com)
 */

(function($) {

$.fn.extend({
    countable: function(givenOptions) {
        return this.each(function() {
            var $this = $(this), interval, prev_char_diff, $el,
                options = $.extend({
                    threshold: 0.5,
                    appendMethod: 'insertAfter', // insertBefore || insertAfter || prependTo || appendTo
                    target: $this, // relative element with which to place the counter
                    startOpacity: 0.25,
                    maxLength: parseInt( $this.attr('maxlength'), 10 ) || 0,
                    maxClassName: 'maxed',
                    className: 'counter',
                    tagName: 'span',
                    submitButton: $(this).closest("form").find("input[type=submit]"),
                    submitDisableClassName: 'disabled',
                    interval: 750,
                    positiveCopy: 'You have {n}&nbsp;characters left.',
                    negativeCopy: 'You are {n}&nbsp;characters over.',
                    fadeDuration: 'normal',
                    defaultText: '' // text to disregard in the character count
                }, givenOptions);

            // create counter element
            $el = $('<'+options.tagName+'/>')
                .html( options.positiveCopy.replace('{n}', '<span class="num"></span>') )
                .addClass( options.className );

            // set initial opacity to 0 if opacity is supported
            if ( $.support.opacity ) $el.css({ opacity: 0 }); // don't set opacity for IE to avoid clear text issues.

            // sppend counter element to the DOM
            $el[options.appendMethod](options.target);

            // hook up events for the input/textarea being monitored
            $this
                .bind('keyup', check)
                .bind('focus blur', function(event) {
                    if ( event.type == 'blur' ) clearInterval( interval );
                    if ( event.type == 'focus' && !interval ) setInterval(check, options.interval);
                });

            // actual function to do the character counting and notification
            function check() {
                var val = $this.val(),
                    length = (val == options.defaultText ? 0 : val.length),
                    percentage_complete = length/options.maxLength,
                    char_diff = options.maxLength - length,
                    opacity;

                // return if we haven't made any progress
                if ( prev_char_diff != undefined && char_diff == prev_char_diff ) return;

                // if counter element is hidden and we are past the given threshold, show it
                if ( $el.is(':hidden') && percentage_complete >= options.threshold )
                    $el.show();
                // if the counter element is visible and we are now under the given threshold, hide it
                if ( $el.is(':visible') && percentage_complete < options.threshold )
                    $el.hide();

                if ( $.support.opacity ) { // don't set opacity for IE to avoid clear type issues.
                    // calculate the correct opacity
                    opacity = options.startOpacity + ((options.threshold - percentage_complete) * ((options.startOpacity * 2) - 2));
                    // animate to the correct opacity
                    $el.stop().fadeTo( options.fadeDuration, percentage_complete >= options.threshold ? opacity : 0 );
                }

                // set the correct copy if under or over the max number of characters
                if ( char_diff >= 0 ) {
                    if ( $el.is( '.'+options.maxClassName ) )
                        $el.html( options.positiveCopy.replace('{n}', '<span class="num"></span>') );

                    if ( $(options.submitButton).is('input[type=submit], button') )
                    {
                      $(options.submitButton).removeClass(options.submitDisableClassName);
                      $(options.submitButton).removeAttr("disabled");
                    }
                } else {
                    if ( !$el.is( '.'+options.maxClassName ) )
                        $el.html( options.negativeCopy.replace('{n}', '<span class="num"></span>') );

                    if ( $(options.submitButton).is('input[type=submit], button') )
                    {
                      $(options.submitButton).addClass(options.submitDisableClassName);
                      $(options.submitButton).attr("disabled", "disabled");
                    }
                }

                // add or remove the max class name
                $el[ (char_diff < 0 ? 'add' : 'remove') + 'Class' ]( options.maxClassName );

                // set the number of characters left or number of characters over the limit
                $el.find('.num').text( Math.abs(char_diff) );

                // make sure the plural is necessary or not
                if ( char_diff == -1 || char_diff == 1 )
                    $el.html( $el.html().replace(/characters\b/, 'character') );
                else
                    $el.html( $el.html().replace(/character\b/, 'characters') );

                prev_char_diff = char_diff;
            };
            // run an initial check
            check();
        });
    }
});

})(jQuery);